cli/module_graph.rs

```diff
@@ -0,0 +1,725 @@
+  maybe_spec: &Option<ModuleSpecifier>,
+  s: S,
+) -> Result<S::Ok, S::Error>
+where
+  S: Serializer,
+{
+  if let Some(spec) = maybe_spec {
+    serialize_module_specifier(spec, s)
+  } else {
+    s.serialize_none()
+  }
+}
+
+#[derive(Debug, Serialize)]
+pub struct ModuleGraph(HashMap<String, ModuleGraphFile>);
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ImportDescriptor {
+  specifier: String,
+  #[serde(serialize_with = "serialize_module_specifier")]
+  resolved_specifier: ModuleSpecifier,
+  // These two fields are for support of @deno-types directive
+  // directly prepending import statement
+  type_directive: Option<String>,
+  #[serde(serialize_with = "serialize_option_module_specifier")]
+  resolved_type_directive: Option<ModuleSpecifier>,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ReferenceDescriptor {
+  specifier: String,
+  #[serde(serialize_with = "serialize_module_specifier")]
+  resolved_specifier: ModuleSpecifier,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ModuleGraphFile {
+  pub specifier: String,
+  pub url: String,
+  pub filename: String,
+  pub imports: Vec<ImportDescriptor>,
+  pub referenced_files: Vec<ReferenceDescriptor>,
+  pub lib_directives: Vec<ReferenceDescriptor>,
+  pub types_directives: Vec<ReferenceDescriptor>,
+  pub type_headers: Vec<ReferenceDescriptor>,
+  pub media_type: i32,
+  pub source_code: String,
+}
+
+type SourceFileFuture =
+  Pin<Box<dyn Future<Output = Result<SourceFile, ErrBox>>>>;
+
+pub struct ModuleGraphLoader {
+  permissions: Permissions,
+  file_fetcher: SourceFileFetcher,
+  maybe_import_map: Option<ImportMap>,
+  pending_downloads: FuturesUnordered<SourceFileFuture>,
+  pub graph: ModuleGraph,
+  is_dyn_import: bool,
+  analyze_dynamic_imports: bool,
+}
+
+impl ModuleGraphLoader {
+  pub fn new(
+    file_fetcher: SourceFileFetcher,
+    maybe_import_map: Option<ImportMap>,
+    permissions: Permissions,
+    is_dyn_import: bool,
+    analyze_dynamic_imports: bool,
+  ) -> Self {
+    Self {
+      file_fetcher,
+      permissions,
+      maybe_import_map,
+      pending_downloads: FuturesUnordered::new(),
+      graph: ModuleGraph(HashMap::new()),
+      is_dyn_import,
+      analyze_dynamic_imports,
+    }
+  }
+
+  pub async fn add_to_graph(
+    &mut self,
+    specifier: &ModuleSpecifier,
+  ) -> Result<(), ErrBox> {
+    self.download_module(specifier.clone(), None)?;
+
+    loop {
+      let load_result = self.pending_downloads.next().await.unwrap();
+      let source_file = load_result?;
+      let spec = ModuleSpecifier::from(source_file.url.clone());
+      self.visit_module(&spec, source_file)?;
+      if self.pending_downloads.is_empty() {
+        break;
+      }
+    }
+
+    Ok(())
+  }
+
+  pub fn build_local_graph<S: BuildHasher>(
+    &mut self,
+    _root_name: &str,
+    source_map: &HashMap<String, String, S>,
+  ) -> Result<(), ErrBox> {
```

Comment:
`build_local_graph()` would also be improved by a triple slash comment.

Why is it generic on `BuildHasher`? Can that not just be left out?

Reply:
`BuildHasher` is used because this method is used by `runtime_compiler` which is also generic over `BuildHasher`. I'm not entirely sure why, but I'll definitely refactor it in follow up PRs.